### ISSUE TITLE
True Lossless

### ISSUE THREAD

**Issue by christian-oreilly** [lina-usc/pylossless#196]
If I am not mistaken, there are some filtering and re-referencing operations that are performed early in the pipeline and the effect of these transformation are saved along with the annotations, making the pipeline not truly lossless which limits the value of its output as the original bidsifyied data + annotations and cleaning procedure. 

In the pipeline saving procedure, we have https://github.com/lina-usc/pylossless/blob/c647153bd39315b94f39a122a2e8629a3c1f08d5/pylossless/pipeline.py#L1129-L1136

This is certain to cause some lossy transformation, for exemple this one https://github.com/lina-usc/pylossless/blob/c647153bd39315b94f39a122a2e8629a3c1f08d5/pylossless/pipeline.py#L1173

In a truly lossless version, the EDF files in the original BIDS and the derivatives should be numerically identical and the annotations along with a the preprocessing procedure (filtering, referencing, etc.) should be saved alongside.

**Comment by scott-huberty** [comment]
@christian-oreilly Reviewer 1 put this under a "miscellaneous comments" section:

> It may be worth exploring whether the pipeline could be implemented without any irreversible modifications. While filtering is necessary for subsequent processing steps (e.g., ICA), one potential approach could be to retain an untouched copy of the original data within the pipeline. This would increase the size of the pipeline object but would preserve both a destructively processed version for analysis and an unaltered version for sharing and reapplying the pipeline.


From the language, it seems like we may be able to resubmit without implementing this, but by this point several people have requested this feature. Do you think that it is worth/feasible to implement before resubmission?

**Comment by christian-oreilly** [comment]
I don't think it is a very complicate thing to implement... All operations would need to be done on the temporary raw object, that can be filtered and what not. On the saving, this raw object should not be saved, only the annotations and the ordered list of non-invertible operations (like filtering) to be performed by the rejection policy... Definitely worth implementing. Feasible before the submission? Not sure. At least, I am a bit overcommitted these days so I am not sure I'll find time to do anything substantial there in the short term.

**Comment by scott-huberty** [comment]
Same here : ( 

But at least we have it on the board, I presume that this will get implemented at some point. 

I'll put this at the bottom of my resubmissions to-do list, and if I have time will take a shot at it.
